python/cudf/cudf/core/resample.py

```diff
@@ -0,0 +1,360 @@
+
+    def agg(self, func):
+        result = super().agg(func)
+        if len(self.grouping.bin_labels) != len(result):
+            index = cudf.Index(
+                self.grouping.bin_labels, name=self.grouping.names[0]
+            )
+            return result._align_to_index(
+                index, how="right", sort=False, allow_non_unique=True
+            )
+        else:
+            return result.sort_index()
+
+    def asfreq(self):
+        return self.obj._align_to_index(
+            self.grouping.bin_labels,
+            how="right",
+            sort=False,
+            allow_non_unique=True,
+        )
+
+    def _scan_fill(self, method: str, limit: int) -> DataFrameOrSeries:
+        # TODO: can this be more efficient?
+
+        # first, compute the outer join between `self.obj` and the `bin_labels`
+        # to get the sampling "gaps":
+        upsampled = self.obj._align_to_index(
+            self.grouping.bin_labels,
+            how="outer",
+            sort=True,
+            allow_non_unique=True,
+        )
+
+        # fill the gaps:
+        filled = upsampled.fillna(method=method)
+
+        # filter the result to only include the values corresponding
+        # to the bin labels:
+        return filled._align_to_index(
+            self.grouping.bin_labels,
+            how="right",
+            sort=False,
+            allow_non_unique=True,
+        )
+
+
+class DataFrameResampler(_Resampler, DataFrameGroupBy):
+    pass
+
+
+class SeriesResampler(_Resampler, SeriesGroupBy):
+    pass
+
+
+class _ResampleGrouping(_Grouping):
+
+    bin_labels: cudf.Index
+
+    def _handle_frequency_grouper(self, by):
+        # if `by` is a time frequency grouper, we bin the key column
+        # using bin intervals specified by `by.freq`, then use *that*
+        # as the groupby key
+        import cudf._lib.labeling
+
+        freq = by.freq
+        label = by.label
+        closed = by.closed
+
+        if isinstance(freq, (cudf.DateOffset, pd.DateOffset)):
+            raise NotImplementedError(
+                "Resampling by DateOffset objects is not yet supported."
+            )
+        if not isinstance(freq, str):
+            raise TypeError(
+                f"Unsupported type for freq: {type(freq).__name__}"
+            )
+        # convert freq to a pd.DateOffset:
+        offset = pd.tseries.frequencies.to_offset(freq)
+
+        if offset.freqstr == "M" or offset.freqstr.startswith("W-"):
+            label = "right" if label is None else label
+            closed = "right" if closed is None else closed
+        else:
+            label = "left" if label is None else label
+            closed = "left" if closed is None else closed
+
+        # determine the key column
+        if by.key is None and by.level is None:
+            # then assume that the key is the index of `self._obj`:
+            if not isinstance(self._obj.index, cudf.DatetimeIndex):
+                raise TypeError(
```

Comment:
Couldn't we remove this instance check/error case and it would still throw almost the same error once it made it down to the check on line 133?

Reply:
Yup - you're right. Fixed!